cpp/net/connection_pool.cc

```diff
@@ -16,45 +17,92 @@ DEFINE_int32(url_fetcher_max_conn_per_host_port, 4,
 namespace cert_trans {
 namespace internal {
 
+ConnectionPool::Connection::Connection(evhtp_connection_t* conn,
+                                       HostPortPair&& other_end)
+    : conn_(CHECK_NOTNULL(conn)), other_end_(move(other_end)) {
+}
+
+
+const HostPortPair& ConnectionPool::Connection::other_end() const {
+  return other_end_;
+}
+
+
 ConnectionPool::ConnectionPool(libevent::Base* base)
     : base_(CHECK_NOTNULL(base)), cleanup_scheduled_(false) {
 }
 
 
-evhttp_connection_unique_ptr ConnectionPool::Get(const URL& url) {
+// static
+evhtp_res ConnectionPool::Connection::ConnectionClosedHook(
+    evhtp_connection_t* conn, void* arg) {
+  CHECK_NOTNULL(conn);
+  CHECK_NOTNULL(arg);
+  VLOG(1) << "Releasing connection " << conn;
+  ConnectionPool::Connection* const c(
+      static_cast<ConnectionPool::Connection*>(arg));
+  CHECK_EQ(conn, c->connection());
+  c->conn_.release();
+  return EVHTP_RES_OK;
+}
+
+
+unique_ptr<ConnectionPool::Connection> ConnectionPool::Get(const URL& url) {
   // TODO(pphaneuf): Add support for other protocols.
   CHECK_EQ(url.Protocol(), "http");
-  const HostPortPair key(url.Host(), url.Port() != 0 ? url.Port() : 80);
+  HostPortPair key(url.Host(), url.Port() != 0 ? url.Port() : 80);
   lock_guard<mutex> lock(lock_);
 
   auto it(conns_.find(key));
+
+  if (it != conns_.end() && !it->second.empty()) {
+    // Do a sweep and remove any dead connections
+    for (auto deque_it(it->second.begin()); deque_it != it->second.end();) {
+      CHECK(*deque_it);
```

Comment:
I think you're trying to do `CHECK_NOTNULL(deque_it->get())` here, right? That might give you a nicer log message, if it ever failed, I think...